package-res/resources/web/pentaho/visual/base/model.js

```diff
@@ -45,37 +73,287 @@ define([
      * @description Creates a base `Model`.
      * @param {pentaho.visual.base.spec.IModel} modelSpec A plain object containing the model specification.
      */
-    var Model = Complex.extend({
-          meta: {
-            id: "pentaho/visual/base",
-            view: "View",
-            isAbstract: true,
-            props: [
-              {
-                name: "width",
-                type: "number",
-                isRequired: true
-              },
-              {
-                name: "height",
-                type: "number",
-                isRequired: true
-              },
-              {
-                name: "isInteractive",
-                type: "boolean",
-                value: true
-              },
-              {
-                name: "data",
-                type: "object",
-                isRequired: true
-              }
-            ]
+    var Model = Complex.extend().implement(EventSource).extend(/** @lends pentaho.visual.base.Model# */{
+        constructor: function(modelSpec) {
+          this.base(modelSpec);
+        },
+
+        /**
+         * Modifies the current selection based on an input filter and on a selection mode.
+         *
+         * This action is the entry point for user-driven modifications of the current selection,
+         * e.g. if the user clicked a bar in a bar chart,
+         * or drew a rectangle over a set of bars in a bar chart.
+         *
+         * The event ["will:select"]{@link pentaho.visual.base.events.WillSelect}
+         * is first emitted and processed.
+         * Its event listeners can be attributed a _priority_,
+         * and can be regarded as operations in a processing pipeline that are allowed to:
+         * - cancel the event,
+         * - mutate the input filter
+         * - mutate the selection mode.
+         *
+         * Afterwards, [_changeSelection]{@link pentaho.visual.base.Model#_changeSelection}
+         * is invoked to compute and update the current selection.
+         *
+         * If the modification of the current selection is successful, the event
+         * ["did:select"]{@link pentaho.visual.base.events.DidSelect} is emitted and processed,
+         * and a [result]{@link pentaho.lang.ActionResult} containing
+         * the updated current selection in the `value` field is returned.
+         *
+         * Any failure (due to an event cancelation, due to listeners throwing exceptions
+         * or due to an invalid selection mode) yields an error [result]{@link pentaho.lang.ActionResult}
+         * and lead to the emission (and processing) of the
+         * ["rejected:select"]{@link pentaho.visual.base.events.RejectedSelect}.
+         * In this case, this method returns a [result]{@link pentaho.lang.ActionResult}
+         * with an exception in the `error` field.
```

Comment:
with an exception in the error field -> "with an error in the `error`" field or "with an `Error` in the `error` field"